Detect and warn plugins that access blocks dependencies without declaring dependencies

ralucaStan opened the pull request:
Fixes WOOPLUG-6043

This PR introduces a mechanism to display console warnings when any of the `wc.X` objects exposed are accessed by a plugin script that didn't declare the corresponding block dependencies in PHP.

# Warnings Mechanism:
1. **Early proxy setup** - PHP outputs an inline script in <head> (before any other scripts run) that wraps `window.wc` in a JavaScript Proxy. This captures the original `wc` object and intercepts all property access.

2. **Property access interception** - When any script accesses a tracked property (e.g., `wc.blocksCheckout`), the Proxy's get handler fires. It identifies the caller by either checking `document.currentScript.src` or parsing a stack trace to find the calling script's URL.

3. **Deferred checking with registry** - Since the proxy runs early but script registration happens later, checks are queued until PHP outputs a second script in the footer containing the registry - a JSON map of script URLs to their handles and declared dependencies.

4. **Dependency validation** - Once the registry loads, queued checks (and any new accesses) compare the caller's declared dependencies against what's required. If the required handle is missing, a console warning identifies the script and the missing dependency.

5. **Warning types - Three scenarios trigger warnings**:
Inline/unknown script - No URL could be identified
Unregistered script - URL not in the registry (not registered via wp_enqueue_script)
Missing dependency - Script is registered but doesn't declare the required dependency



# Testing 
Make sure you test in Chrome and in Firefox as well                      
Prerequisites: 
- Enable debug mode: define( 'WP_DEBUG', true ); in wp-config.php
- Have a page with the Checkout block (or Cart/Mini-Cart)
- Verify detection is enabled: see info message: `[WooCommerce] Dependency detection enabled...`

## Plugins correctly declaring dependencies - no warnings
- Install WooCommerce Points and Rewards or WooCommerce subscriptions (these use the IntegrationInterface)
- Open the browser console & notice that the dependency initialization started : [WooCommerce] Dependency detection enabled. Warnings will be shown for scripts that access wc.* globals without proper dependencies.
- Notice there are no dependecy warnings for scripts coming from these extensions

## Plugins not declaring dependencies

- Install the wc-dependecy-test plugin ([download link](https://drive.google.com/file/d/18wLogyeWqy19tIw3KUBAlOMuNuUYOv3z/view?usp=sharing) ) & activate

Notice the console messages:

### Registered scripts with missing dependecy
The plugin registers the `bad-extension-test` script without dependencies. The script accesses `wc.blocksCheckout`;
Notice the warning below:
<img width="818" height="128" alt="Screenshot 2025-12-02 at 19 15 49" src="https://github.com/user-attachments/assets/9d9e21cd-00a9-4dcf-aa04-9baa92e05bf1" />


### Unregistered scripts: 
The plugin programatically adds a script to the `<head>` which calls `wc.priceFormat`; look after `/wp-content/plugins/wc-dependency-test/test-unregistered-script.js`.
Notice the warning below: 
```
[wooCommercel Unregistered script "test-unregistered-script.js" accessed wc.priceFormat. This script should be registered with w_enqueue_script() and declare "wc-price- format" as a dependency.
```
<img width="539" height="84" alt="Screenshot 2025-12-18 at 18 04 38" src="https://github.com/user-attachments/assets/e1eca743-c131-45b9-ba27-d3f79403b7ae" />

### Scripts that can't be identified
Sometimes we can't identify a caller script, like in this case, where `setTimeout` is used. 
Why: `document.currentScript` is null because the script has already finished executing. The fallback is parsing the error stack trace to find the .js URL. But setTimeout callbacks run asynchronously from the browser's event loop, so the stack shows `at getCallerScriptUrl (inline-script:50:17)`. 
<img width="822" height="62" alt="Screenshot 2025-12-02 at 19 16 50" src="https://github.com/user-attachments/assets/25fae7cd-6fb0-4d10-9a29-6f11d4f2e5b0" />


## Test calls made before the registry exists are being made
The mechanism collects the calls made to the proxy and checks the dependencies after the registry is outputed in the footer.
- Install the wc-dependecy-test plugin ([download link](https://drive.google.com/file/d/18wLogyeWqy19tIw3KUBAlOMuNuUYOv3z/view?usp=sharing) ) & activate
- Add a product to cart and go to the checkout page
- Notice the console messages:

<img width="815" height="100" alt="Screenshot 2025-12-02 at 19 28 49" src="https://github.com/user-attachments/assets/8615548f-95bf-40e4-a9be-f8cb7a2d422b" />

opr committed:
Update local pickup selector to immediately reflect selection in client

opr committed:
Add changelog

opr committed:
Update props of local pickup Slot

opr committed:
Update tests to check selectedInClient parameter

opr committed:
Fix lint error

opr committed:
Make isSelectedInClient an optional param

opr committed:
Pass selected option to callback to allow callback decide what to do

opr committed:
Merge branch 'trunk' into wooplug-5811-checkout-block-pickup-location-details-renders-awkwardly

ralucaStan committed:
Dependency injection

ralucaStan committed:
Remove WC_DEBUG_DEPENDENCIES and checks for additional dependencies

ralucaStan committed:
Improve detection of async calls

ralucaStan committed:
Revert "Pass selected option to callback to allow callback decide what to do"

This reverts commit 25ca5e7e55db594df206f47c48b163a2a6160774.

ralucaStan committed:
Revert "Add changelog"

This reverts commit dfbd239f74e380fa981be2d86108ef50da6e9bb7.

ralucaStan committed:
Revert "Update props of local pickup Slot"

This reverts commit ee4a987db10fb532e53350e50557ac064b2a0989.

ralucaStan committed:
Revert "Update tests to check selectedInClient parameter"

This reverts commit 36d625d9ed96c2c16932fa9db536242e5b41d803.

ralucaStan committed:
Revert "Make isSelectedInClient an optional param"

This reverts commit 7d870a3eb7c97f99aec01bc948f0758ad0d5e7bb.

ralucaStan committed:
Revert "Pass selected option to callback to allow callback decide what to do"

This reverts commit 25ca5e7e55db594df206f47c48b163a2a6160774.

ralucaStan committed:
Revert commits from intial branch

ralucaStan committed:
Revert changes from initial branch

opr commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,415 @@
+	 */
+	public function output_early_proxy_setup(): void {
+		?>
+		<script id="wc-dependency-detection-early">
```
What do you think of moving this script to a `.js` file and loading it here rather than doing it inline like so, that way we can maybe run tests against the script and get the benefit of our IDEs formatting it and linting it etc.

ralucaStan replied:
The initial approach was using a JS file that was enqueued, but there were some timing issues.
We can move the JS code to a file, read its contents, and output it in the script tag. Is this what you have in mind?

opr replied:
> We can move the JS code to a file, read its contents, and output it in the script tag. Is this what you have in mind?

Exactly, I understood the issue of enqueueing and thought this would be a good idea, separating into its own JS file makes it easier to search/read/test imo.

ralucaStan committed:
Add a separate JS file for front-end detection logic

ralucaStan committed:
Add all scripts that are exposed on the window.wc object

ralucaStan committed:
Clean up code:
Remove the registering logic for the initially used JS
Rename the file for the current JS logic file
Remove unused AssetApi class usage

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan committed:
Align WooCommerce URL detection

ralucaStan committed:
Make PHP is the source of truth for the global-to-handle mapping

ralucaStan committed:
Merge branch 'spike/dependency-detection' of github.com:woocommerce/woocommerce into spike/dependency-detection

ralucaStan committed:
Remove proxy set trap and don't use Reflect.set

ralucaStan committed:
Rename and spacing

ralucaStan committed:
Only run logic on pages with the checkout blocks

ralucaStan committed:
Improve proxy creation and registry output
Now hooks directly to wp_head and wp_print_footer_scripts
Avoid registry output if proxy didn't work

ralucaStan committed:
Comment logic for script warnings

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan committed:
clean js detection:
- remove proxyEnabled, we keep it on the php side
- rename for more clarity

ralucaStan committed:
Remove WordPress internal scripts from the registry
As we output this registry in the front-end we don't need all the registered scripts from WordPress, only the ones coming from plugins

ralucaStan committed:
When we can't identify the calling script from the stack trace, it will warn immediately as an "inline or unknown script" without waiting for or checking the registry.

ralucaStan committed:
improve logic to detect caller url in stack error

ralucaStan committed:
Improve detection to prevent recursice proxy get calls
the wc globals call inside them other exposed wc packages because the webpack externals configuration maps all @woocommerce/* imports to window.wc.*.
example: priceFormat imports @woocommerce/settings which becomes window.wc.wcSettings at runtime, triggering our proxy.

ralucaStan committed:
Run dependency detection only when debugging is enabled

ralucaStan committed:
Simplify recursion prevention in dependency detection
Instead of parsing stack traces to identify recursive calls, we prevent them from running at all.
The isChecking flag blocks nested proxy calls while a dependency check is in progress, eliminating the need to
detect recursion by analyzing stack traces.

ralucaStan committed:
Changelog

ralucaStan committed:
fix PHPStan warnings

ralucaStan committed:
clean comments

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan committed:
PhPStan baseline file

ralucaStan committed:
Add tests for the dependency js file
This implied exporting parts of the js and using webpack to build the file
PHP now reads from the build folder

ralucaStan committed:
Add PHP tests

ralucaStan committed:
Fix lint issues

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan committed:
Fix linting

ralucaStan committed:
use wp_parse_url to normalize links

ralucaStan committed:
keep WC_GLOBAL_EXPORTS variable name when minifying

ralucaStan committed:
Add ventor to JS pattern

ralucaStan committed:
Follow unit test guidelines

ralucaStan committed:
Ensure the proxy guard is reset in case caller detection fails

ralucaStan committed:
Move the registry function on the wc global to avoid conflicts

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+			'__WC_GLOBAL_EXPORTS_PLACEHOLDER__',
+			$mapping_json,
+			$script_content
+		);
```
fixed in 76f67618eb0bae11d13a37c7c735c66925dc2037; the function was moved under `wc.wcUpdateDependencyRegistry`

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 * Initialize hooks.
+	 */
+	public function init(): void {
+		// Only run when debugging is enabled.
```
This is the desired behaviour, because we don't need devs to enable a separate flag to show them the warnings. This should work out of the box for developers who have WP_DEBUG enabled

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 *
+	 * @var array<string, string>
+	 */
+	private const WC_GLOBAL_EXPORTS = array(
```
This is just a warning mechanism. removing keys from the `wc` global object would be a breaking change and communicated as such when the time comes.

ralucaStan committed:
Use template literals instead of string concatenation

ralucaStan committed:
Update test file

ralucaStan commented on plugins/woocommerce/tests/php/src/Blocks/DependencyDetectionTest.php:
```diff
@@ -0,0 +1,289 @@
+/**
+ * Unit tests for the DependencyDetection class.
+ */
+class DependencyDetectionTest extends \WP_UnitTestCase {
```
fixed in 7f1e963786194321f8beb20c1223728930ddee2f

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+ * Pattern to identify WooCommerce core scripts (which we should skip).
+ * Matches /plugins/woocommerce/(client|assets|build)/ but NOT /plugins/woocommerce-subscriptions/ etc.
+ */
+const WC_CORE_SCRIPT_PATTERN =
```
The pattern only targets scripts coming from the WooCommerce pattern, which should be ignored. It's only targeting extensions or scripts outside of WooCommerce

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 */
+	public function init(): void {
+		// Only run when debugging is enabled.
+		if ( ! defined( 'WP_DEBUG' ) || ! WP_DEBUG ) {
```
Not needed, we don't want this behaviour in production

ralucaStan committed:
Follow guidelines in test file

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 */
+	public function output_early_proxy_setup(): void {
+		// Only run on pages that have the tracked blocks.
+		if ( ! $this->page_has_tracked_blocks() ) {
```
Extensions shouldn't and have no reason to do that interception

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,209 @@
+		return value;
+	}
+
+	return new Proxy( target, { get: __wcProxyGet } );
```
fixed in adca5956f993d62b62c01583b1db1dd8c5d172e9

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 * Based on wcDepMap and wcHandleMap in client/blocks/bin/webpack-helpers.js.
+	 *
+	 * @var array<string, string>
+	 */
```
> they should remain in the array but marked as deprecated for at least one major version before removal.

This is just a warning mechanism. There will be a transition period for plugins to deal with these warnings. If globals won't be exposed on `wc` anymore, that will be communicated accordingly, and the warnings dependencies will be modified as well.

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 *
+	 * @var array<string>
+	 */
+	private const TRACKED_BLOCKS = array(
```
> Extensions that rely on being able to set or override window.wc properties early may break if they expect to control the window.wc object initialization.

Extensions shouldn't have to do that. They only access the window.wc object keys

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.ts:
```diff
@@ -0,0 +1,174 @@
+		checkDependency
+	);
+
+	// Define window.wc as a getter/setter to maintain the proxy.
```
>  Extensions defining their own property descriptors on window.wc should set configurable: true and check if a descriptor already exists before defining their own.

Extensions should only access the wc globals.

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	/**
+	 * Initialize hooks.
+	 */
+	public function init(): void {
```
> Extensions should not rely on the presence or absence of dependency detection. Don't check for wcUpdateDependencyRegistry to determine if detection is active.

Extensions access `wc.X` without declaring dependencies. That is because usually they are already exposed, and so their plugins work. This is just a warnings mechanism which should inform plugin developers that they need to also declare the dependency. It's not to be accessed by them

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.ts:
```diff
@@ -0,0 +1,175 @@
+		configurable: true,
+		enumerable: true,
+	} );
+
```
No external code should update the registry.

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/test/utils.test.ts:
Should this be a TS file, and does it need the `.test` suffix?

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+ * Matches /plugins/woocommerce/(client|assets|build)/ but NOT /plugins/woocommerce-subscriptions/ etc.
+ */
+const WC_CORE_SCRIPT_PATTERN =
+	/\/plugins\/woocommerce\/(client|assets|build|vendor)\//;
```
Apparently this can be customised so we should try to get the actual plugins directory from the server rather than hardcoding it

```
// wp-config.php - security through obscurity or corporate conventions
define('WP_PLUGIN_DIR', '/var/www/html/wp-content/extensions');
define('WP_PLUGIN_URL', 'https://example.com/wp-content/extensions');

// Or the whole content directory
define('WP_CONTENT_DIR', '/var/www/html/app');
define('WP_CONTENT_URL', 'https://example.com/app');
```

ralucaStan replied:
d02467e65d3cc824f26a222f1b1646c84ba6bd69

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+export function getFilename( url ) {
+	if ( ! url ) return 'unknown';
+
+	const filename = url.split( '/' ).pop().split( '?' )[ 0 ].split( '#' )[ 0 ];
```
I think we should be more careful here, `pop()` can return undefined on an empty array, which would throw when calling `.split()` on it.

ralucaStan replied:
cbbf6b6e67c0c3ad82b3e84bbb8156a03abfb3ab

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
I feel this should also be a TS file, could you please update and ensure the file is typed as well as possible? I think it would be OK to `@ts-expect-error` things that would be out of scope to fix.

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.ts:
I feel this should also be a TS file, could you please update and ensure the file is typed as well as possible? I think it would be OK to `@ts-expect-error` things that would be out of scope to fix.

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.js:
```diff
@@ -0,0 +1,175 @@
+	 * @return {string|null} The caller script URL or null.
+	 */
+	function getCallerScriptUrl() {
+		if ( document.currentScript && document.currentScript.src ) {
```
Let's check `.src` is a string rather than just truthy here, before accessing `.replace` below

ralucaStan replied:
1ea4ee0e041743b694fbcc7ac1030adfdcc0b1d6

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+ * @return {string} The filename or 'unknown'.
+ */
+export function getFilename( url ) {
+	if ( ! url ) return 'unknown';
```
Nit pick but I would like it if we could keep the curly braces here even for one line ifs.

ralucaStan replied:
cbbf6b6e67c0c3ad82b3e84bbb8156a03abfb3ab

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+}
+
+/**
+ * Check if a stack line should be skipped (internal code).
```
It would be helpful if this could be more descriptive, why are we skipping?

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+ */
+export function extractJsUrl( line ) {
+	const match = line.match( /(https?:\/\/[^\s)]+\.js)(?:[?:#]|$)/ );
+	return match ? match[ 1 ] : null;
```
`line` isn't type-checked before calling `.match()` I think it's worth adding a guard for consistency with `getFilename`?

ralucaStan replied:
369edc9821916dbc0a1c26850d0e036a34089da2

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+ * @return {string|null} The caller URL or null if not found.
+ */
+export function parseStackForCallerUrl( stack, currentPage ) {
+	if ( ! stack ) return null;
```
Braces would be great here

ralucaStan replied:
cbbf6b6e67c0c3ad82b3e84bbb8156a03abfb3ab

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+export function parseStackForCallerUrl( stack, currentPage ) {
+	if ( ! stack ) return null;
+
+	const lines = stack.split( '\n' );
```
Let's double check `stack` is a string here, because this method is exported we have no guarantee it's going to always receive a string

ralucaStan replied:
99cb58a5fe66a4a60eb7631a5887f5b593b9ec6c

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+		};
+	}
+
+	const scriptInfo = scriptRegistry[ callerUrl ];
```
Let's double check `scriptRegistry` is an object before accessing a property of it

ralucaStan replied:
a818b5966fbc0eee051b106301f567cdbedd0995

opr commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+	const scriptInfo = scriptRegistry[ callerUrl ];
+
+	// Case 2: Unregistered script.
+	if ( ! scriptInfo ) {
```
I think harder checks are needed on `scriptInfo` here since it's assumed to be an object with at least `deps` (array)

ralucaStan replied:
a818b5966fbc0eee051b106301f567cdbedd0995

opr reviewed (changes requested):
Thanks for working on this! I have some code suggestions, and also I'm wondering if Firefox/Chrome/Safari all format their stack traces in the same way, claude seems to think that Firefox formats them differently, can we verify and if necessary, add a test for this?

I'd prefer it if all new files we add would be in TS with correct types, please would you work on this, and add `@ts-expect-error` if necessary for stuff that's out of scope for this PR?

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan committed:
Merge branch 'spike/dependency-detection' of github.com:woocommerce/woocommerce into spike/dependency-detection

ralucaStan committed:
Convert dependency files to TS

ralucaStan committed:
Improve getFilename

ralucaStan committed:
Handle malformed registry:
Registry entry with missing deps property
Registry entry with non-array deps
Registry itself being null/undefined/not an object

ralucaStan committed:
Ensure stack is a string and handle gracefully

ralucaStan committed:
Ensure currentScript.src is a string

ralucaStan committed:
Ensure line is a string

ralucaStan committed:
fix webpack for dependecy warning assets

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 *
+	 * @var array<string>
+	 */
+	private const TRACKED_BLOCKS = array(
```
👍

ralucaStan committed:
Improve comments on types and shouldSkipLine goal

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.ts:
```diff
@@ -0,0 +1,204 @@
+	);
+
+	// Define window.wc as a getter/setter to maintain the proxy.
+	Object.defineProperty( window, 'wc', {
```
Extensions should only read from window.wc.

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.ts:
```diff
@@ -0,0 +1,204 @@
+		configurable: true,
+		enumerable: true,
+	} );
+
```
extensions shouldn't write to wc, and the key is not likely to be used by extensions.

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,324 @@
+	 *
+	 * @var array<string, string>
+	 */
+	private const WC_GLOBAL_EXPORTS = array(
```
we will document the Proxy in the release. the removal of the wc keys will be announced.

ralucaStan committed:
Add different stack format handling

ralucaStan committed:
Document that exposed function is internal

ralucaStan committed:
Account for the plugins directory not being /plugins

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan committed:
Improve substring checks and account for malformed src

ralucaStan committed:
Ensure the detection js file is the blocks folder, and not in the build
so that it's not ignored on live envs

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.js:
```diff
@@ -0,0 +1,193 @@
+ * @param {Object}   target             - The object to wrap.
+ * @param {Object}   wcGlobalExports    - Map of wc.* properties to required handles.
+ * @param {Function} getCallerScriptUrl - Function to get the caller script URL.
+ * @param {Function} checkDependency    - Function to check and warn about dependencies.
```
> Additionally, the proxy introduces timing-dependent behavior where accessing wc.* properties before the registry loads versus after produces different warning behavior, potentially affecting extension initialization order.
The detection logic collects all calls and checks them when the regitry loads.

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/utils.ts:
```diff
@@ -0,0 +1,267 @@
+ */
+export interface ScriptInfo {
+	handle: string;
+	deps: WcDependencyHandle[];
```
In the future, the properties won't be removed without a deprecation warning

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/dependency-detection/index.ts:
```diff
@@ -0,0 +1,206 @@
+			}
+			pendingChecks = [];
+		};
+
```
We might consider a woocommerce_dependency_detection_console_message filter in a follow up PR. @opr for awareness.

ralucaStan committed:
Handle wp_json_encode() failure for registry

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,347 @@
+		// Only output registry if the proxy was set up earlier.
+		// This avoids the duplicate page_has_tracked_blocks() check and ensures
+		// we don't output a registry without a proxy to consume it.
+		if ( ! $this->proxy_output ) {
+			return;
+		}
+
+		// Build the script registry mapping URLs to handles and dependencies.
+		$script_registry = $this->build_script_registry();
+		$registry_json   = \wp_json_encode( $script_registry );
+
+		// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped -- JSON is safely encoded by wp_json_encode.
+		echo '<script id="wc-dependency-detection-registry">if(typeof window.wc.wcUpdateDependencyRegistry==="function"){window.wc.wcUpdateDependencyRegistry(' . $registry_json . ');}</script>' . "\n";
```
1ac3372d564240dd7c9dfc69e9c8bdf3edc55da9

ralucaStan committed:
Add @since 10.5.0 annotations to public methods

ralucaStan commented on plugins/woocommerce/src/Blocks/DependencyDetection.php:
```diff
@@ -0,0 +1,357 @@
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		$this->init();
+	}
```
This is not true, the init is not called twice.
The pattern is:
Classes that call init() in constructor → Bootstrap just calls $container->get()
Classes that don't call init() in constructor → Bootstrap calls $container->get()->init()

ralucaStan committed:
Merge branch 'trunk' into spike/dependency-detection

opr committed:
Merge branch 'trunk' into spike/dependency-detection

opr committed:
Add null checks and tests

opr committed:
lint fix

opr reviewed (approved):
thanks for the changes Raluca, I've added a couple more checks where we seemed to be unsafely accessing properties on unknown types, but I think it's good now. I tested on Checkout and the example plugin worked, I also modified the example plugin to work on Cart and I also saw the warnings there, thanks for working on this.
